Mathlib/Algebra/PresentedMonoid/Basic.lean

```diff
@@ -0,0 +1,248 @@
+  rw [map_mul, hx, hy, map_mul]
+
+@[simp]
+theorem toMonoid.of {x : α} : (PresentedMonoid.toMonoid f h) (PresentedMonoid.of rels x) =
+    f x := by
+  simp only [PresentedMonoid.of, PresentedMonoid.lift, toMonoid, MonoidHom.coe_mk, OneHom.coe_mk,
+    FreeMonoid.lift_eval_of]
+
+end ToMonoid
+
+@[ext]
+theorem ext {M : Type*} [Monoid M] (rels : FreeMonoid α → FreeMonoid α → Prop)
+    {φ ψ : PresentedMonoid rels →* M} (hx : ∀ (x : α), φ (.of rels x) = ψ (.of rels x)) :
+    φ = ψ := by
+  ext a
+  induction' a with b
+  induction b
+  · rw [← one_def, map_one, map_one]
+  · rename_i x
+    exact hx x
+  rename_i x y hx hy
+  rw [← mul_mk, map_mul, map_mul, hx, hy]
+
+section Isomorphism
+variable {β : Type*} (e : α ≃ β) (rels : FreeMonoid α → FreeMonoid α → Prop)
+
+/--given an isomorphism between two types, convert a relation function to take in the iso type -/
+def rels_iso : FreeMonoid β → FreeMonoid β → Prop :=
+  fun a b => rels ((FreeMonoid.congr_iso e).invFun a) ((FreeMonoid.congr_iso e).invFun b)
+
+private theorem iso_helper : (StepsTo rels ⇒ StepsTo (rels_iso e rels))
+    (FreeMonoid.congr_iso e).toFun (FreeMonoid.congr_iso e).toFun := by
+  intro x y h
+  induction h with
+  | basic rxy =>
+    apply StepsTo.basic
+    unfold rels_iso
+    simp only [MulEquiv.toEquiv_eq_coe, Equiv.toFun_as_coe, EquivLike.coe_coe, Equiv.invFun_as_coe,
+      EquivLike.coe_symm_apply_apply]
+    exact rxy
+  | left _ ih =>
+    rw [(FreeMonoid.congr_iso e).map_mul', (FreeMonoid.congr_iso e).map_mul']
+    exact StepsTo.left ih
+  | right _ ih =>
+    rw [(FreeMonoid.congr_iso e).map_mul', (FreeMonoid.congr_iso e).map_mul']
+    exact StepsTo.right ih
+
+private theorem iso_helper_inv : (StepsTo (rels_iso e rels) ⇒ StepsTo rels)
+    (FreeMonoid.congr_iso e).invFun (FreeMonoid.congr_iso e).invFun := by
+  intro x y h
+  have H : ∀ x y, ((FreeMonoid.congr_iso e).invFun (x*y)) =
+      ((FreeMonoid.congr_iso e).invFun x) * ((FreeMonoid.congr_iso e).invFun y) := by
+    intro x y
+    rcases (FreeMonoid.congr_iso e).surjective x with ⟨_, ha⟩
+    rcases (FreeMonoid.congr_iso e).surjective y with ⟨_, hb⟩
+    rw [← ha, ← hb]
+    simp
+  induction h with
+  | basic rxy =>
+    unfold rels_iso at rxy
+    exact StepsTo.basic rxy
+  | left _ ih =>
+    rw [H, H]
+    exact StepsTo.left ih
+  | right _ ih =>
+    rw [H, H]
+    exact StepsTo.right ih
+
+/-- forward direction of the bijection for the isomorphism -/
+private def iso_function : PresentedMonoid rels → PresentedMonoid (rels_iso e rels) :=
+  Quot.map ((FreeMonoid.congr_iso e).toFun) (iso_helper e rels)
+
+/-- reverse direction of the bijection for the isomorphism -/
+private def iso_function_inv : PresentedMonoid (rels_iso e rels) → PresentedMonoid rels :=
+  Quot.map ((FreeMonoid.congr_iso e).invFun) (iso_helper_inv e rels)
+
+/-- Presented groups of isomorphic types are isomorphic. -/
+def equivPresentedMonoid : PresentedMonoid rels ≃* PresentedMonoid (rels_iso e rels) := by
```

Comment:
This is a general result that If `e : M →* N` is surjective then `(c.comap e).Quotient ≃* c.Quotient` with `c : Con N`.

Reply:
implemented in the latest PR!